src/librustdoc/html/highlight.rs

```diff
@@ -6,15 +6,32 @@
 //! Use the `render_with_highlighting` to highlight some rust code.
 
 use crate::html::escape::Escape;
+use crate::html::render::Context;
 
-use std::fmt::Display;
+use std::fmt::{Display, Write};
 use std::iter::Peekable;
 
 use rustc_lexer::{LiteralKind, TokenKind};
 use rustc_span::edition::Edition;
 use rustc_span::symbol::Symbol;
 
-use super::format::Buffer;
+use super::format::{self, Buffer};
+use super::render::{LightSpan, LinkFromSrc};
+
+/// This type is needed in case we want to render links on items to allow to go to their definition.
+crate struct ContextInfo<'a, 'b, 'c> {
+    crate context: &'a Context<'b>,
+    /// This represents the "lo" bytes of the current file we're rendering. To get a [`Span`] from
+    /// it, you just need to add add your current byte position in the string and its length (to get
+    /// the "hi" part).
+    ///
+    /// This is used to create a [`LightSpan`] which is then used as an index in the `span_map` in
+    /// order to retrieve the definition's [`Span`] (which is used to generate the URL).
+    crate file_span_lo: u32,
+    /// This field is used to know "how far" from the top of the directory we are to link to either
+    /// documentation pages or other source pages.
+    crate root_path: &'c str,
```

Comment:
Isn't this what Context.current is for? Why did you introduce a new variable? Reusing current would let you get rid of the new parameter to href I think.

Reply:
No, current contains a vec of modules, the point here is to generate ".." for each element of `current`. I took it from `layout::Page` in `sources.rs`.